src/v2/dry_run.rs

```diff
@@ -0,0 +1,1250 @@
+use super::{
+    generated::{
+        self, account_transaction_payload, dry_run_error_response, dry_run_request,
+        DryRunInvokeInstance, DryRunMintToAccount, DryRunSignature, DryRunStateOperation,
+    },
+    AccountIdentifier, IntoBlockIdentifier,
+};
+use crate::{
+    types::{
+        smart_contracts::{ContractContext, InstanceInfo, ReturnValue},
+        AccountInfo, AccountTransactionDetails, RejectReason,
+    },
+    v2::{generated::DryRunStateQuery, Require},
+};
+use concordium_base::{
+    base::{Energy, ProtocolVersion},
+    common::types::{CredentialIndex, KeyIndex, Timestamp},
+    contracts_common::{AccountAddress, Amount, ContractAddress},
+    hashes::BlockHash,
+    smart_contracts::ContractTraceElement,
+    transactions::{EncodedPayload, PayloadLike},
+};
+use futures::*;
+
+mod shared_receiver {
+    use futures::{stream::Stream, StreamExt};
+    use tokio::{
+        sync::{mpsc, oneshot},
+        task::JoinHandle,
+    };
+
+    /// A `SharedReceiver` wraps an underlying stream so that multiple clients
+    /// can queue to receive items from the stream.
+    pub struct SharedReceiver<S>
+    where
+        S: Stream, {
+        senders: mpsc::UnboundedSender<oneshot::Sender<S::Item>>,
+        task:    JoinHandle<()>,
+    }
+
+    impl<S: Stream> Drop for SharedReceiver<S> {
+        fn drop(&mut self) { self.task.abort(); }
+    }
+
+    impl<S: Stream + Unpin + Send + 'static> SharedReceiver<S>
+    where
+        S::Item: Send,
+    {
+        /// Construct a new shared receiver. This spawns a background task that
+        /// pairs waiters with incoming stream items.
+        pub fn new(stream: S) -> Self {
+            let (senders, rec_senders) = mpsc::unbounded_channel::<oneshot::Sender<S::Item>>();
+            let task = tokio::task::spawn(async {
+                let mut zipped = stream.zip(tokio_stream::wrappers::UnboundedReceiverStream::from(
+                    rec_senders,
+                ));
+                while let Some((item, sender)) = zipped.next().await {
+                    let _ = sender.send(item);
+                }
+            });
+            SharedReceiver { senders, task }
+        }
+
+        /// Claim the next item from the stream when it becomes available.
+        /// Returns `None` if the stream is already closed. Otherwise returns a
+        /// [`oneshot::Receiver`] that can be `await`ed to retrieve the item.
```

Comment:
Why not propagate the SendError instead?

Reply:
That needlessly exposes an implementation detail. Morever, the `SendError` would merely wrap the sender end of the new oneshot channel, the other end of which will have already been dropped, and so is completely useless to any caller.